### PR TITLE
feat!: split code between new com package and core

### PR DESCRIPTION
**Pull request by barak007**
This PR extract the communication from `@wixc3/engine-core` into `@wixc3/engine-com`

Why:
* This is the most early tech debt in core
* Separation of concerns 
* ease of documentation 

Tasks:

- [x] split files
- [x] split tests
- [ ] list breaking changes
- [ ] remove todos
- [ ] ensure types and commons are good
- [x] replace usages of exports in other packages
- [ ] SERVICE_CONFIG?
- [ ] extenal typo

Breaking changes: 
- runEngineApp
- export split 
- no internal index?
- utils split

This is a major version

### CODE REVIEW

File: packages/com/src/com/initializers/iframe.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
-import type { WindowHost } from '../types';
+import type { WindowHost } from '../../types';
 import type { Communication } from '../communication';
 import { isIframe } from '../helpers';
-import { injectScript } from '../../helpers';
+import { injectScript } from '../inject-script';
```
Comment (thehulke) [inline packages/com/src/com/initializers/iframe.ts:4]: Why removing it from `helpers`?

File: packages/com/src/com/initializers/window.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
-import type { WindowHost } from '../types';
+import type { WindowHost } from '../../types';
 import { isIframe } from '../helpers';
-import { injectScript } from '../../helpers';
+import { injectScript } from '../inject-script';
```
Comment (thehulke) [inline packages/com/src/com/initializers/window.ts:3]: same question

File: packages/com/test/communication.spec.ts [hunk 3]
```diff
@@ -1,12 +1,8 @@
...
-import {
-    Communication,
-    Environment,
-    declareComEmitter,
-    iframeInitializer,
-    deferredIframeInitializer,
-} from '@wixc3/engine-core';
+import { Environment } from '@wixc3/engine-core';
+import { Communication, declareComEmitter, iframeInitializer, deferredIframeInitializer } from '@wixc3/engine-com';
+
```
Comment (thehulke) [inline packages/com/test/communication.spec.ts:5]: can remove line

File: packages/com/test/node/env-dependencies.types.spec.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import { Environment, Feature } from '@wixc3/engine-core';
+import { Service, AsyncApi, EnvironmentInstanceToken } from '@wixc3/engine-com';
+import { typeCheck } from '../type-check';
+import type { EQUAL } from 'typescript-type-utils';
+
+/** type check only test */
+describe.skip('ENV dependencies', () => {
```
Comment (thehulke) [inline packages/com/test/node/env-dependencies.types.spec.ts:7]: can we remove `.skip`?

File: packages/com/test/type-check.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export function typeCheck<U extends true, T extends (...args: U[]) => U>(_fn: T): void {
```
Comment (thehulke) [inline packages/com/test/type-check.ts:1]: Can we use it in a `common` way? 
looks like we have this file multiple times

File: packages/engineer/src/tsconfig.json [hunk 6]
```diff
@@ -4,20 +4,11 @@
...
-    },
-    {
-      "path": "../../runtime-node/src"
-    }
+    { "path": "../../core/src" },
+    { "path": "../../com/src" },
+    { "path": "../../core-node/src" },
+    { "path": "../../scripts/src" },
+    { "path": "../../dashboard/src" },
+    { "path": "../../runtime-node/src" }
```
Comment (thehulke) [inline packages/engineer/src/tsconfig.json:12]: Should sync linters in IDE

File: packages/runtime-node/src/tsconfig.json [hunk 7]
```diff
@@ -4,12 +4,5 @@
...
   },
-  "references": [
-    {
-      "path": "../../core/src"
-    },
-    {
-      "path": "../../core-node/src"
-    }
-  ]
+  "references": [{ "path": "../../core/src" }, { "path": "../../com/src" }, { "path": "../../core-node/src" }]
```
Comment (thehulke) [inline packages/runtime-node/src/tsconfig.json:7]: IDE Linter sync

File: packages/scripts/src/create-entrypoint.ts [hunk 8]
```diff
@@ -7,6 +7,9 @@ import type { IFeatureDefinition } from './types';
 const { stringify } = JSON;
 const topLevelConfigLoaderPath = require.resolve('./top-level-config-loader');
 export const LOADED_FEATURE_MODULES_NAMESPACE = '_engine_';
+//!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+//TODO: !!!!!!!!!!!!!!!!!!!!!!!! publicPath, resolvedContexts should be passed to COM.config
+//!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (thehulke) [inline packages/scripts/src/create-entrypoint.ts:12]: uhm

File: packages/test-kit/src/create-browser-provider.ts [hunk 9]
```diff
@@ -12,7 +12,7 @@ export function createBrowserProvider(options?: playwright.LaunchOptions) {
             return page;
         },
         async dispose() {
-            if (browser && browser.isConnected) {
+            if (browser && browser.isConnected()) {
```
Comment (thehulke) [inline packages/test-kit/src/create-browser-provider.ts:15]:
```suggestion
            if (browser?.isConnected()) {
```

File: packages/test-kit/src/with-feature.ts [hunk 10]
```diff
@@ -125,7 +125,7 @@ let executableApp: IExecutableApplication;
 if (typeof after !== 'undefined') {
     after('close browser, if open', async function () {
         this.timeout(20_000);
-        if (browser && browser.isConnected) {
+        if (browser && browser.isConnected()) {
```
Comment (thehulke) [inline packages/test-kit/src/with-feature.ts:128]:
```suggestion
        if (browser?.isConnected()) {
```

**Review by thehulke — COMMENTED** [review]
from what i understand now it looks nice overall 
some minor comments though

File: packages/com/src/com/initializers/iframe.ts (hunk 1 above)
Comment (barak007) [reply packages/com/src/com/initializers/iframe.ts:4]: give it a proper place

File: packages/com/src/com/initializers/window.ts (hunk 2 above)
Comment (barak007) [reply packages/com/src/com/initializers/window.ts:3]: same answer

File: packages/com/test/node/env-dependencies.types.spec.ts (hunk 4 above)
Comment (barak007) [reply packages/com/test/node/env-dependencies.types.spec.ts:7]: there are here only for type check

File: packages/engineer/src/tsconfig.json (hunk 6 above)
Comment (barak007) [reply packages/engineer/src/tsconfig.json:12]: ?

File: packages/runtime-node/src/tsconfig.json (hunk 7 above)
Comment (barak007) [reply packages/runtime-node/src/tsconfig.json:7]: ?

File: packages/com/src/communication.feature.ts [hunk 11]
```diff
@@ -1,14 +1,8 @@
...
 import { LoggerService } from './com/logger-service';
-import type { Target } from './com/types';
-import { Config } from './entities/config';
-import { AllEnvironments, Universal } from './entities/env';
-import { Feature } from './entities/feature';
-import { Value } from './entities/value';
-import { Slot } from './entities/slot';
-import { RUN_OPTIONS, ENGINE } from './symbols';
-import { LoggerTransport, LogLevel } from './types';
+import { AllEnvironments, Universal, Config, RUN_OPTIONS, ENGINE, Feature, Slot, Value } from '@wixc3/engine-core';
```
Comment (AviVahl) [inline packages/com/src/communication.feature.ts:4]: circularity

File: packages/com/src/service.ts [hunk 12]
```diff
@@ -1,10 +1,16 @@
...
+    RuntimeEngine,
+    CREATE_RUNTIME,
+    REGISTER_VALUE,
+    EnvVisibility,
+    AllEnvironments,
+    Environment,
+    normEnvVisibility,
+    Universal,
+    FeatureOutput,
+} from '@wixc3/engine-core';
```
Comment (AviVahl) [inline packages/com/src/service.ts:13]: circularity

File: packages/com/src/service.ts [hunk 13]
```diff
@@ -66,19 +72,19 @@ export class Service<
             }
 
             // eslint-disable-next-line @typescript-eslint/no-unsafe-return
-            return inputValue || this.getApiProxy(runtimeEngine, serviceKey);
+            return inputValue || (this.getApiProxy(runtimeEngine, serviceKey) as any);
```
Comment (AviVahl) [inline packages/com/src/service.ts:75]: as unknown?

**Comment by barak007** [comment]
outdated 😢

**Pull request closed without merging**